### PR TITLE
Add host functions for bigint

### PR DESCRIPTION
**Pull request by jayz22**
### What

[TODO: Short statement about what is changing.]

### Why

[TODO: Why this change is being made. Include any context required to understand the why.]

### CODE REVIEW

File: stellar-contract-env-host/src/host.rs [hunk 1]
```diff
@@ -1332,99 +1336,150 @@ impl CheckedEnv for Host {
...
+        let res = self.visit_obj(x, |a: &BigInt| {
+            Ok(self.visit_obj(y, |b: &BigInt| Ok(a.gcd(b)))?)
+        })?;
+        Ok(self.add_host_object(res)?.into())
     }
 
     fn bigint_lcm(&self, x: Object, y: Object) -> Result<Object, HostError> {
-        todo!()
+        let res = self.visit_obj(x, |a: &BigInt| {
+            Ok(self.visit_obj(y, |b: &BigInt| Ok(a.lcm(b)))?)
```
Comment (jonjove) [inline stellar-contract-env-host/src/host.rs:1496]: Also looks like this handles 0 correctly, but we should be sure to test this (same thing about opening an issue).

File: stellar-contract-env-common/src/env.rs [hunk 2]
```diff
@@ -175,26 +175,35 @@ macro_rules! call_macro_with_all_host_functions {
...
-                {"H", fn bigint_gcd(x:Object, y:Object) -> Object}
-                {"I", fn bigint_lcm(x:Object, y:Object) -> Object}
-                {"J", fn bigint_pow(x:Object, y:RawVal) -> Object}
-                {"K", fn bigint_pow_mod(p:Object, q:Object, m:Object) -> Object}
-                {"L", fn bigint_sqrt(x:Object) -> Object}
-                {"M", fn bigint_bits(x:Object) -> u64}
+                /// Creates a BigInt from a binary array `x` and i32 sign `s`.
+                /// Bytes in big-endian order. Sign is interpreted: -1 as negative, 0 as zero, 1 as positive
+                /// If sign is 0, then the input bytes are ignored and will return a BigInt of 0.
+                {"3", fn bigint_from_bytes_be(x:Object, s:RawVal) -> Object}
```
Comment (jonjove) [inline stellar-contract-env-common/src/env.rs:181]: Do we really need these functions? They aren't mentioned in CAP-0051, and I'm not entirely sure how they would be used right now.

Part of my concern is that these interfaces seem a bit like foot-guns. For example, should we ignore `x` if `s == 0` or should we trap? I can see that the library handles the case where `x` is empty but `s != 0` by making the result 0, but should we trap in that case? It's all pretty unclear to me.

Also based on other conventions you've been using I would expect `s` to be `u32`.

In summary, there are a lot of questions here. I think you should drop this commit for now and we should work out these details in the CAP first, instead of trying to work them out in code where it is much harder and more time consuming.

**Review by jonjove — CHANGES_REQUESTED** [review]

File: stellar-contract-env-host/src/host.rs [hunk 3]
```diff
@@ -597,7 +598,18 @@ impl Host {
                     HostObject::U64(u) => Ok(ScObject::U64(*u)),
                     HostObject::I64(i) => Ok(ScObject::I64(*i)),
                     HostObject::Bin(b) => Ok(ScObject::Binary(b.clone().try_into()?)),
-                    HostObject::BigInt(_) => todo!(),
+                    HostObject::BigInt(bi) => {
+                        let (sign, data) = bi.to_bytes_be();
+                        match sign {
+                            Sign::Minus => {
```
[the remark this answers is not in the document]
Comment (jayz22) [reply stellar-contract-env-host/src/host.rs:608]: My bad. Fixed.

File: stellar-contract-env-host/src/host.rs [hunk 4]
```diff
@@ -1332,99 +1336,150 @@ impl CheckedEnv for Host {
...
+        let res = self.visit_obj(x, |a: &BigInt| {
+            Ok(self.visit_obj(y, |b: &BigInt| Ok(a.mul(b)))?)
+        })?;
+        Ok(self.add_host_object(res)?.into())
     }
 
     fn bigint_div(&self, x: Object, y: Object) -> Result<Object, HostError> {
-        todo!()
+        let res = self.visit_obj(x, |a: &BigInt| {
+            Ok(self.visit_obj(y, |b: &BigInt| Ok(a.div(b)))?)
```
[the remark this answers is not in the document]
Comment (jayz22) [reply stellar-contract-env-host/src/host.rs]: Fixed.

File: stellar-contract-env-host/src/host.rs [hunk 5]
```diff
@@ -1332,99 +1336,150 @@ impl CheckedEnv for Host {
...
+        })?;
+        Ok(self.add_host_object(res)?.into())
     }
 
     fn bigint_rem(&self, x: Object, y: Object) -> Result<Object, HostError> {
-        todo!()
+        let res = self.visit_obj(x, |a: &BigInt| {
+            Ok(self.visit_obj(y, |b: &BigInt| Ok(a.rem(b)))?)
```
[the remark this answers is not in the document]
Comment (jayz22) [reply stellar-contract-env-host/src/host.rs]: Fixed.

File: stellar-contract-env-host/src/host.rs [hunk 6]
```diff
@@ -1332,99 +1336,150 @@ impl CheckedEnv for Host {
...
     fn bigint_not(&self, x: Object) -> Result<Object, HostError> {
-        todo!()
+        let res = self.visit_obj(x, |a: &BigInt| Ok(a.not()))?;
+        Ok(self.add_host_object(res)?.into())
     }
 
     fn bigint_gcd(&self, x: Object, y: Object) -> Result<Object, HostError> {
-        todo!()
+        let res = self.visit_obj(x, |a: &BigInt| {
+            Ok(self.visit_obj(y, |b: &BigInt| Ok(a.gcd(b)))?)
```
[the remark this answers is not in the document]
Comment (jayz22) [reply stellar-contract-env-host/src/host.rs]: Test now covers it.

File: stellar-contract-env-host/src/host.rs (hunk 1 above)
Comment (jayz22) [reply stellar-contract-env-host/src/host.rs]: Tests now covers it.

**Comment by jayz22** [comment]
Tests have been added. Also buffed up a lot of the error checking which should make the functions more robust.

File: stellar-contract-env-host/src/test.rs [hunk 7]
```diff
@@ -1173,3 +1168,197 @@ fn binary_xdr_roundtrip() -> Result<(), HostError> {
     assert_eq!(host.obj_cmp(obj.into(), obj_back.into())?, 0);
     Ok(())
 }
+
+#[test]
+fn bigint_tests() -> Result<(), HostError> {
+    let host = Host::default();
+    let a: u64 = 2374340;
+    let b: i64 = -438730;
+    // let bi_0: BigInt = 0_u32.into();
```
Comment (jonjove) [inline stellar-contract-env-host/src/test.rs:1177]: Nitpick: why are these commented out?

File: stellar-contract-env-host/src/test.rs [hunk 8]
```diff
@@ -1173,3 +1168,197 @@ fn binary_xdr_roundtrip() -> Result<(), HostError> {
...
+        let obj_res = host.bigint_gcd(obj_a, obj_b)?;
+        let obj_ref = host.bigint_from_i64(10)?;
+        assert_eq!(host.obj_cmp(obj_res.into(), obj_ref.into())?, 0);
+        // gcd by 0 is self
+        let obj_res = host.bigint_gcd(obj_a, obj_0)?;
+        assert_eq!(host.obj_cmp(obj_res.into(), obj_a.into())?, 0);
+    }
+    // lcm
+    {
+        // let bi_c = bi_a.lcm(&bi_b);
```
Comment (jonjove) [inline stellar-contract-env-host/src/test.rs:1303]: Nitpick: I assume you meant to remove all of these comments (in lcm, pow, pow_mod, sqrt, and bits)?

File: stellar-contract-env-host/src/test.rs [hunk 9]
```diff
@@ -1173,3 +1168,197 @@ fn binary_xdr_roundtrip() -> Result<(), HostError> {
...
+        // println!("{:?}", bi_c) // 104169418820
+        let obj_res = host.bigint_lcm(obj_a, obj_b)?;
+        let obj_ref = host.bigint_from_i64(104169418820)?;
+        assert_eq!(host.obj_cmp(obj_res.into(), obj_ref.into())?, 0);
+        // lcm by 0 is 0
+        let obj_res = host.bigint_lcm(obj_a, obj_0)?;
+        assert_eq!(host.obj_cmp(obj_res.into(), obj_0.into())?, 0);
```
Comment (jonjove) [inline stellar-contract-env-host/src/test.rs:1310]: You should probably also check that lcm(0,0)=0.

File: stellar-contract-env-host/src/host.rs [hunk 10]
```diff
@@ -1342,27 +1342,57 @@ impl CheckedEnv for Host {
     }
 
     fn bigint_add(&self, x: Object, y: Object) -> Result<Object, HostError> {
-        let res = self.visit_obj(x, |a: &BigInt| self.visit_obj(y, |b: &BigInt| Ok(a.add(b))))?;
+        let res = self.visit_obj(x, |a: &BigInt| {
+            self.visit_obj(y, |b: &BigInt| {
+                a.checked_add(b)
```
[the remark this answers is not in the document]
Comment (jayz22) [reply stellar-contract-env-host/src/host.rs:1370]: I know all they do is wrapping result in an `Option`. I think it's more favorable to call the checked version, if one is offered. The [documentation](https://docs.rs/num-bigint/latest/num_bigint/struct.BigUint.html#impl-CheckedAdd) mentions the possibility of overflow, in which case `None` will be returned. This just gives us an extra layer of safety without any added perf cost. If for whatever reason the operation does fail, we will trap the VM properly (with an error code) instead of just crashing.

File: stellar-contract-env-host/src/test.rs (hunk 7 above)
Comment (jayz22) [reply stellar-contract-env-host/src/test.rs:1177]: These are convenience code to generate the result reference. I think it's convenient to have in case one may change the value `a` and `b` and need to regenerate the references.

File: stellar-contract-env-host/src/host.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (jonjove) [reply stellar-contract-env-host/src/host.rs:1370]: But they don't _actually_ do any overflow checking. One might argue that the overflow checking could be added in a future version. And this is true, but in that case it would be a protocol change anyway and we would need to deploy that carefully.

File: stellar-contract-env-host/src/host.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jayz22) [reply stellar-contract-env-host/src/host.rs:610]: Oh, you meant these brackets? Rust analyzer automatically inserts them.

File: stellar-contract-env-host/src/test.rs (hunk 7 above)
Comment (jonjove) [reply stellar-contract-env-host/src/test.rs:1177]: I don't understand. What result reference? Whatever the result reference is, why would you need to regenerate?

Even if this is convenience code for some purpose that I don't yet understand, there's no indication that it is convenience code for whatever that purpose is. As such, it's very unlikely that a maintainer would be able to make use of it as currently presented.

File: stellar-contract-env-host/src/host.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonjove) [reply stellar-contract-env-host/src/host.rs:610]: Ah, it's just reaching the line length limit. Nothing to do here, never mind.

**Review by jonjove — APPROVED** [review]

File: stellar-contract-env-host/src/host.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (jayz22) [reply stellar-contract-env-host/src/host.rs]: I just think if they provided one (and specifically mentioned overflow being a possibility), we should use it in case some lower level dependency panics in ways we didn't expect it. This gives us more control over error emission. 
I un-did all the `checked_` ops. Since there isn't one for every function we call `num-bigint` (e.g. pow), so we are not fully protected either way. And it does make the code a bit cleaner.

File: stellar-contract-env-host/src/test.rs (hunk 7 above)
Comment (jayz22) [reply stellar-contract-env-host/src/test.rs]: Remove them.

File: stellar-contract-env-host/src/host.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (jonjove) [reply stellar-contract-env-host/src/host.rs]: > I just think if they provided one (and specifically mentioned overflow being a possibility), we should use it in case some lower level dependency panics in ways we didn't expect it.

Let's turn to the code and see if this makes sense. Consider `checked_add` which has the implementation
```rust
    #[inline]
    pub fn checked_add(&self, v: &BigInt) -> Option<BigInt> {
        Some(self + v)
    }
```
What would happen if `self + v` were to panic? This function would panic, of course. The `Some` doesn't protect you from anything at all. `Some` _isn't_ error handling in its own right, you need to explicitly provide the error handling. Because `self + v` doesn't return a `Result` or some other type with error handling capabilities, the only thing it can do is panic on error. You mostly cannot catch panics in rust (see https://doc.rust-lang.org/std/panic/fn.catch_unwind.html for the exception, but it's not reliable). But even if you could, this code isn't trying to catch the panics either.

In summary, the returned `Option` is given you a false sense of security. And, again, if a later version added some error checking then that would be a protocol change anyway; we would not want to get that error checking automatically.

Am I making sense here?

[the remark this answers is not in the document]
Comment (jayz22) [reply stellar-contract-env-host/src/host.rs]: Yeah that makes sense. In that case, the right thing to do is to find and check for those conditions ourselves and call the "unchecked" version directly. Sorry I got confused about the panic part.

**Comment by jonjove** [comment]
@jayz22 I forgot to post this earlier when I merged, but I have a bit of meta-feedback on this PR:

1. Force-pushing isn’t part of the workflow in this repository, but I can see you did it in this PR to rebase. You should merge instead in this repository.

2. Many of the changes in “Clean up some bad Rust code, e.g. unnecessary Ok()” have nothing to do with bigint. The title of this PR is “Add host functions for bigint”, so making unrelated changes just makes things more confusing and harder to review (because I’m in the bigint context, but now I have to move into the context for all the other unrelated changes). At the very least, do such changes in a separate commit although a separate PR is probably better.
